src/org/opensolaris/opengrok/search/context/LineHighlight.java

```diff
@@ -0,0 +1,208 @@
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+import org.opensolaris.opengrok.web.HtmlConsts;
+import org.opensolaris.opengrok.web.Util;
+
+/**
+ * Represents a collection of metadata related to highlighting a single line
+ * of code.
+ */
+public class LineHighlight {
+
+    private final int lineno;
+    private List<PhraseHighlight> markups;
+    /** Offset of elided left part */
+    private int lelide;
+    /** Offset of elide right part */
+    private int relide;
+
+    private boolean didLelide;
+    private boolean didRelide;
+
+    public LineHighlight(int lineno) {
+        if (lineno < 0) {
+            throw new IllegalArgumentException("`lineno' cannot be negative");
+        }
+        this.lineno = lineno;
+    }
+
+    /**
+     * Gets the number of markups
+     * @return zero or greater
+     */
+    public int countMarkups() {
+        return markups == null ? 0 : markups.size();
+    }
+
+    /**
+     * Gets the highlight at the specified position.
+     * @param i index of element to return
+     * @return defined instance
+     */
+    public PhraseHighlight getMarkup(int i) {
+        return markups.get(i);
+    }
+
+    /**
+     * Sort and condense overlapping markup highlights.
+     */
+    public void condenseMarkups() {
+        if (markups == null) {
+            return;
+        }
+
+        markups.sort(PhraseHighlightComparator.INSTANCE);
+        // Condense instances if there is overlap.
+        for (int i = 0; i + 1 < markups.size(); ++i) {
+            PhraseHighlight phi0 = markups.get(i);
+            PhraseHighlight phi1 = markups.get(i + 1);
+            if (phi0.overlaps(phi1)) {
+                phi0 = phi0.merge(phi1);
+                markups.set(i, phi0);
+                markups.remove(i + 1);
+                --i;
+            }
+        }
+    }
+
+    /**
+     * Adds the specified highlight.
+     * @param phi a defined instance
+     */
+    public void addMarkup(PhraseHighlight phi) {
+        if (phi == null) {
+            throw new IllegalArgumentException("`phi' is null");
+        }
+        if (markups == null) {
+            markups = new ArrayList<>();
+        }
+        markups.add(phi);
+    }
+
+    /**
+     * @return the lineno
+     */
+    public int getLineno() {
+        return lineno;
+    }
+
+    /**
+     * Gets the left elide value.
+     * @return zero or greater
+     */
+    public int getLelide() {
+        return lelide;
+    }
+
+    /**
+     * Sets the left elide value.
+     */
+    public void setLelide(int value) {
+        if (value < 0) {
+            throw new IllegalArgumentException("`value' is negative");
+        }
+        this.lelide = value;
+    }
+
+    /**
+     * Gets the right elide value.
+     * @return zero or greater
+     */
+    public int getRelide() {
+        return relide;
+    }
+
+    /**
+     * Sets the right elide value.
+     */
+    public void setRelide(int value) {
```

Comment:
I am wondering if we should move the project towards scala or kotlin to make the data objects less verbose

Reply:
come on, less typing ;)
(but then one needs to get ready for some code readability changes - for good or for bad)